packages/backend-storage/src/access_builder.ts

```diff
@@ -0,0 +1,111 @@
+import {
+  AuthRoleName,
+  ConstructFactory,
+  ConstructFactoryGetInstanceProps,
+  ResourceAccessAcceptor,
+  ResourceAccessAcceptorFactory,
+  ResourceProvider,
+} from '@aws-amplify/plugin-types';
+
+export type StorageAction = 'read' | 'write' | 'delete';
+
+export type StorageAccessDefinition = {
+  getResourceAccessAcceptor: (
+    getInstanceProps: ConstructFactoryGetInstanceProps
+  ) => ResourceAccessAcceptor;
+  /**
+   * Actions to grant to this role on a specific prefix
+   */
+  actions: StorageAction[];
+  /**
+   * The value that will be substituted into the resource string in place of the {owner} token
+   */
+  ownerPlaceholderSubstitution: string;
+};
+
+export type StorageAccessBuilder = {
+  to: (...actions: StorageAction[]) => StorageAccessDefinition;
+};
+
+/**
+ * !EXPERIMENTAL!
+ *
+ * Resource access patterns are under active development and are subject to breaking changes.
+ * Do not use in production.
+ */
+export type RoleAccessBuilder = {
+  authenticated: StorageAccessBuilder;
+  guest: StorageAccessBuilder;
+  owner: StorageAccessBuilder;
+  resource: (
+    other: ConstructFactory<ResourceProvider & ResourceAccessAcceptorFactory>
+  ) => StorageAccessBuilder;
+};
+
+export const storageAccessBuilder: RoleAccessBuilder = {
+  authenticated: {
+    to: (...actions) => ({
+      getResourceAccessAcceptor: getAuthRoleResourceAccessAcceptor,
+      actions,
+      ownerPlaceholderSubstitution: '*',
+    }),
+  },
+  guest: {
+    to: (...actions) => ({
+      getResourceAccessAcceptor: getUnauthRoleResourceAccessAcceptor,
+      actions,
+      ownerPlaceholderSubstitution: '*',
+    }),
+  },
+  owner: {
+    to: (...actions) => ({
+      getResourceAccessAcceptor: getAuthRoleResourceAccessAcceptor,
+      actions,
+      ownerPlaceholderSubstitution: '${cognito-identity.amazon.com:sub}',
+    }),
+  },
+  resource: (other) => ({
+    to: (...actions) => ({
+      getResourceAccessAcceptor: (getInstanceProps) =>
+        other.getInstance(getInstanceProps).getResourceAccessAcceptor(),
+      actions,
+      ownerPlaceholderSubstitution: '*',
+    }),
+  }),
```

Comment:
what does `resource` mean here and is `*` correct substitution here?

Reply:
The resource type access doesn't have a cogntio id because those tokens come directly from sts (or however lambda gets tokens) rather than going through cognito